cleanup_switzerland/models/database_cleanup.py

```diff
@@ -0,0 +1,132 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    Copyright (C) 2017 Compassion CH (http://www.compassion.ch)
+#    Releasing children from poverty in Jesus' name
+#    @author: Emanuel Cino <ecino@compassion.ch>
+#
+#    The licence is in the file __manifest__.py
+#
+##############################################################################
+import logging
+
+from datetime import date
+
+from dateutil.relativedelta import relativedelta
+
+from odoo import api, models
+from psycopg2 import sql, ProgrammingError
+
+
+_logger = logging.getLogger(__name__)
+
+
+class DatabaseCleanup(models.AbstractModel):
+    _name = 'database.cleanup.switzerland'
+
+    ##########################################################################
+    #                             FIELDS METHODS                             #
+    ##########################################################################
+    @api.model
+    def cleanup(self):
+        """
+        We use this method called in a CRON in order to clean our database.
+        :return:
+        """
+        self.clean_mail_message()
+        # Clean emails that are not valid (created by incoming messages)
+        self.env['mail.mail'].search([
+            ('state', '=', 'outgoing'),
+            '|', ('author_id', '=', False),
+            ('author_id.user_ids', '=', False)
+        ]).unlink()
+        return True
+
+    def clean_mail_message(self):
+        # Clean messages not related to an existing object
+        _logger.info("Database cleanup for Compassion Switzerland started")
+        self.env.cr.execute(
+            "DELETE FROM mail_message WHERE model IS NULL"
+        )
+        models = self.env['ir.model'].search([]).mapped('model')
+        for model in models:
+            table = model.replace('.', '_')
+            try:
+                self.env.cr.execute(sql.SQL(
+                    "DELETE FROM mail_message WHERE model=%s AND NOT EXISTS ("
+                    "    SELECT 1 FROM {} WHERE id = res_id);"
+                ).format(sql.Identifier(table)),
+                                    [model])
+                # Commit the cleanup
+                count = self.env.cr.rowcount
+                self.env.cr.commit()  # pylint: disable=invalid-commit
+                _logger.info(
+                    '{} mail_message for invalid {} objects removed'.format(
+                        count, model)
+                )
+            except ProgrammingError:
+                self.env.cr.rollback()
+
+        # Delete old messages for old objects not useful
+        today = date.today()
+        limit = today - relativedelta(years=1)
```

Comment:
For donations certificate maybe the accounting team need to see more than just one year in the past?

Reply:
This is only cleaning the messages, meaning the history of the changes of status for instance. We will still be able to see the invoices, the accounting entries and everything needed for donations certificate.